### DISCUSSION TITLE
USB MIDI key press inversion on Step48

### DISCUSSION THREAD

**Discussion by omersiar** [rsta2/circle#509] [Q&A]
I was trying to isolate a bug that was introduced somewhere between Step47 and Step48, it may also possible to have this bug previously but it was at least reported as [rare](https://github.com/probonopd/MiniDexed/pull/755#issuecomment-2480125541) Step48 makes it more reproducible in different setups (bunch of different RPis, variety of USB devices).

I see `#ifdef XHCI_DEBUG` is available to us, but it seems my setup (RPi 3A+, loglevel=4) does not provide anything useful.  

Any ideas on how to gather more information?

Bug lies somewhere between USB - MIDI - GadgetMode.

**Comment by omersiar** [comment]
Sorry forgot the add what is the issue,

With a MiniDexed device, too many repeated  USB midi keyboard key press makes noteoff and noteon messages inverted so the keydown event actually getting processed as noteoff midi message and vice versa.

https://github.com/probonopd/MiniDexed/pull/755

**Comment by rsta2** [comment]
Can you try to revert the commit ae00d9d8 for a test? This is the only one between Step47 and 48, which affected the DWHCI USB host controller driver. But this has nothing to do with gadget mode. I think there have been no modification to the USB MIDI gadget driver in this time interval.

If this doesn't help you could only do a `git bisect` to find the commit, which caused this issue.

**Comment by rsta2** [comment]
Have you been able to test this with the reverted commit?

**Reply by omersiar** [reply]
Not me but @soyersoyer able to bisect it

https://github.com/probonopd/MiniDexed/pull/755#issuecomment-2567132377

**Reply by rsta2** [reply]
Thanks @soyersoyer, I will see, what I can do about this.

**Comment by rsta2** [comment]
The affected commit has been reverted on the *develop* branch. It still can be activated with the system option `USE_NAK_USB_FIX`, which shouldn't be enabled for MiniDexed. This modification will be part of the next release.

**Comment by rsta2** [comment]
This problem has been fixed in Circle 49.
